srcpkgs/pgbackrest/template

```diff
@@ -0,0 +1,20 @@
+# Template file for 'pgbackrest'
+pkgname=pgbackrest
+version=2.15.1
+revision=1
+archs="i686 x86_64* aarch64*"
```

Comment:
why limit archs?

Reply:
See https://github.com/void-linux/void-packages/pull/13411#issuecomment-516325679

Reply:
The error doesn't seem to be related to perl, though?

Reply:
It's weird because the error suggests something wrong with the `mtune` param which is not in the cross-compile profile: 
`cc1: error: unrecognized -mtune target: generic`

https://github.com/void-linux/void-packages/blob/98b2e82f9546b0ee8b950066e6399e11c3b2c927/common/cross-profiles/armv6l.sh#L6

I'll dig deeper in case I will be able to find something ...

Reply:
In case you won't succeed: archs should be [blacklisted](https://github.com/void-linux/void-packages/blob/d80cf81922ea8b5e8fdfea58ac046b670f50dae5/srcpkgs/cparser/template#L16) rather than whitelisted, because people build Void for not officially supported archs: [ppc](https://github.com/void-ppc/void-packages), [risc-v](https://github.com/void-linux/void-packages/pull/13207), mips, i686-musl and more. Additionally, this allows to leave note for reason it doesn't build.

Reply:
@Chocimier why should I submit a new package that I maintain and add `archs` that I did not even test whether they can be build successfully? 

I've seen this `archs="i686 x86_64* aarch64*"` basically everywhere in the `void-packages` repo. I never knew it's wrong.

Reply:
Primary reason is that semantics of `archs` is "this is bound to listed architectures and makes no sense for others, because it is written in assembler, repacked from binary deb or so". For semantics of "incidentally not building on aarch64" one use `broken` inside `case` linked above.

Secondly, you do not want to demand additional work from @q66 of figuring out if it is truely not suitable for ppc, or just not tested. Or in general, to be future-proof.

There is also same reason for which you do not mark `archs="aarch64-musl aarch64 armv6hf-musl armv6hf armv6l-musl armv6l armv7hf-musl armv7hf armv7l-musl armv7l i686-musl i686 x86_64 x86_64-musl"` when everything goes smooth just to make sure no one try to build on Itanium.

There are not so many `archs` set. After excluding `="i686* x86_64* ppc*"` which are mostly legitimate, or misspelled `nocross`, there are about 50.

Personally, in this case, I think that releasing stuff that may work is more useful that not releasing because it may not work.

Reply:
pgbackrest builds just fine on ppc. For whatever reason, the cflags from x86_64 build-profile are leaking into your build, though - that seems to be the cause of the problem

Reply:
I figured it out. There is nothing in `xbps-src` leaking the flags in. I just tried building it for `aarch64` on my `ppc64le` host. I got this in `config.log`

```
CC='aarch64-linux-gnu-gcc'
CFLAGS='-fstack-clash-protection -D_FORTIFY_SOURCE=2 -O2 -pipe -march=armv8-a  -g -I/usr/aarch64-linux-gnu/usr/include'
CINCLUDE='-I. -I../libc'
CLIBRARY='  -fstack-clash-protection -D_FORTIFY_SOURCE=2 -mcpu=powerpc64le -mtune=power9 -maltivec -mabi=elfv2 -O2 -pipe   -g -D_FILE_OFFSET_BITS=64 -DLARGE_FILE_SUPPORT64    -I/usr/lib/perl5/core_perl/CORE  -I/usr/aarch64-linux-gnu/usr/include/libxml2'
```

Note how `CLIBRARY` contains flags for `ppc64le`. This is not set anywhere in the environment - if you grep `CLIBRARY` in the source tree, you will get this:

```
configure.ac:AC_SUBST(CLIBRARY, "`perl -MExtUtils::Embed -e ccopts`")
configure.ac:AC_SUBST(CLIBRARY, "$CLIBRARY `xml2-config --cflags`")
```

And that's the source of our problem:

```
q66@banana: /home/q66$ perl -MExtUtils::Embed -e ccopts
  -fstack-clash-protection -D_FORTIFY_SOURCE=2 -mcpu=powerpc64le -mtune=power9 -maltivec -mabi=elfv2 -O2 -pipe   -g -D_FILE_OFFSET_BITS=64 -DLARGE_FILE_SUPPORT64    -I/usr/lib/perl5/core_perl/CORE
```

It's using host binary of perl to get the compiler flags perl was built with, assuming a non-cross-compiling environment. This is obviously bogus, it should not be doing that. This only includes arch flags and hardening, which is already set by the `CFLAGS` we input in from `xbps-src`. Worse problem is `LDFLAGS`, which is also retrieved using a similar way, see the `LIBS` variable in the same file

```
q66@banana: /home/q66$ perl -MExtUtils::Embed -e ldopts
-Wl,-E -Wl,-z,relro -Wl,-z,now -Wl,--as-needed     -pthread  -L/usr/lib/perl5/core_perl/CORE -lperl -lm -lcrypt -lpthread -ldl
```

This contains arch specific, but meaningful, linkage information, that is not passed from `xbps-src`.

Reply:
The immediate solution that involves least work is making it `nocross=yes` (but do not set any `archs`). Any successful cross build is a coincidence and built with incorrect flags as it is. Making it cross-build correctly would involve patching the build system.

Reply:
Thanks, will do so tomorrow!